### PR TITLE
Refactor apply_transform

### PR DESCRIPTION
**Pull request by roman-kh**

### CODE REVIEW

File: batchflow/batch.py [hunk 1]
```diff
@@ -582,168 +505,87 @@ def apply_transform(self, func, *args, **kwargs):
...
-        tr_res = func(*_args, **kwargs)
-
-        if dst is None:
-            pass
-        elif isinstance(dst, str):
-            setattr(self, dst, tr_res)
-        else:
-            dst[:] = tr_res
-        return self
+        if p is None or p == 1:
```
Comment (SergeyTsimfer) [inline batchflow/batch.py:586]: Special treatment for named expression `P` in `inbatch_parallel` is very hard to anticipate

File: batchflow/batch.py [hunk 2]
```diff
@@ -1089,3 +932,31 @@ def dump(self, *args, dst=None, fmt=None, components=None, **kwargs):
     def save(self, *args, **kwargs):
         """ Save batch data to a file (an alias for dump method)"""
         return self.dump(*args, **kwargs)
+
+    @apply_parallel_
+    def to_array(self, comp, dtype=None, channels='last'):
```
Comment (SergeyTsimfer) [inline batchflow/batch.py:937]: Maybe `float32` as default `dtype`? This action already has a strong flavour of ML, so it sounds reasonable

File: batchflow/batch.py [hunk 3]
```diff
@@ -582,168 +505,87 @@ def apply_transform(self, func, *args, **kwargs):
...
-            apply_transform(B.some_method, p=.5)
+            apply_parallel(make_masks_fn, src='images', dst='masks')
+            apply_parallel(apply_mask, src=('images', 'masks'), dst='images')
+            apply_parallel(rotate, src=['images', 'masks'], dst=['images', 'masks'], p=.2, componentwise=True)
+            apply_parallel(MyBatch.some_static_method, p=.5)
+            apply_parallel(B.some_method, p=.5)
         """
-        kwargs_full = {**self.transform_defaults, **kwargs}
-        target, init, post, _ = [kwargs_full.pop(keyname) for keyname in ['target', 'init', 'post', 'all']]
+        kwargs_full = {**self.apply_defaults, **kwargs}
```
[the remark this answers is not in the document]
Comment (roman-kh) [reply batchflow/batch.py:533]: Meta class is used for decorated methods only whereas apply_parallel can also be called directly.

File: batchflow/batch.py [hunk 4]
```diff
@@ -582,168 +505,87 @@ def apply_transform(self, func, *args, **kwargs):
...
 
-        parallel = inbatch_parallel(init=init, post=post, target=target)
-        transform = parallel(type(self)._apply_transform)
-        return transform(self, func, *args, **kwargs_full)
+        src, dst = [kwargs_full.pop(keyname, None) for keyname in ['src', 'dst']]
```
[the remark this answers is not in the document]
Comment (roman-kh) [reply batchflow/batch.py:535]: applying default would become cumbersome

File: batchflow/batch.py [hunk 5]
```diff
@@ -1089,3 +932,31 @@ def dump(self, *args, dst=None, fmt=None, components=None, **kwargs):
     def save(self, *args, **kwargs):
         """ Save batch data to a file (an alias for dump method)"""
         return self.dump(*args, **kwargs)
+
+    @apply_parallel_
```
[the remark this answers is not in the document]
Comment (roman-kh) [reply batchflow/batch.py:936]: no

File: batchflow/batch.py (hunk 1 above)
Comment (roman-kh) [reply batchflow/batch.py:586]: I don't think I understand your comment

File: batchflow/batch.py [hunk 6]
```diff
@@ -582,168 +505,87 @@ def apply_transform(self, func, *args, **kwargs):
...
-                src_attr = [getattr(self, component)[self.get_pos(None, component, ix)] for component in src]
-            else:
-                pos = self.get_pos(None, dst, ix)
-                src_attr = (src[pos],)
-            _args = tuple([*src_attr, *args])
+            init = src
 
-        if p is None or np.random.binomial(1, p):
-            return func(*_args, **kwargs)
+        if p is not None and not isinstance(p, P):
```
[the remark this answers is not in the document]
Comment (roman-kh) [reply batchflow/batch.py]: yes

**Pull request merged**